### PR TITLE
[WIP] Hardware Tessellation on DX9

### PR DESCRIPTION
**Pull request by xebra**
[Edit]
I implemented HW tess to DX9 but still something incomplete.
Basically, tessellation is correct with bezier test in pspgl. But  it can't compute normals because temp register is not enough.

Temp register error
log

Bezier with normals
menu_00007
Bezier with texture without normals
menu_00008
Bezier with lines
menu_00009

Spline without normals
menu_00011
Spline with lines
menu_00010

### CODE REVIEW

**Comment by xebra** [comment]
If these issues does not fix, this implementation is not useful.
If anyone know to fix these issues, please tell me any information.

Thanks in advance.

**Comment by hrydgard** [comment]
Right, I'm not going to merge until it works better. Still, interesting to see what can be done with good old DX9, though it's really pushing it...

**Comment by xebra** [comment]
Yeah, I'm just curious about the potential of old graphics api.

**Review by finansar222 — APPROVED** [review]

**Comment by xebra** [comment]
Almost fixed. FF4CC and Pursuit Force seems perfect.
Remaining issue is about registers for compute normals.

**Comment by hrydgard** [comment]
Nice work :)

**Comment by xebra** [comment]
Thanks :)

**Comment by unknownbrackets** [comment]
Does it help to move more things into functions?  Maybe the compiler isn't smart enough to clobber `bernderiv` by the time it gets to `du = tess_sample(_pos, bernderiv_u)` or something, without it being in a separate scope...?

Or maybe if we calculated `pos_r`, `pos_l`, and `du` first, and then calculated `pos_d`, `pos_u`, and `dv` second?  That way there'd be less total floats needed at any given time (assuming it doesn't optimize the code by reordering it.)

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, older shader compilers (and some newer ones...) sometimes need a lot of help with register allocation. 

This needs a rebase now after merging #9465 but it should be simple.

**Comment by xebra** [comment]
Hmm, I already tried a lot of trial and error trying to fix it. However, I got an idea when I'm thinking about to improve software tessellation. It's to move weight calculation to the CPU. I realized that weight calculation needs to calculate only ONCE. So we can reuse precalculated weights if it's in a cache. Also if tess factor is same like `tess_u == tess_v`, we can reuse precalculated u weights instead calculate v weights. ( `weights_v = weights_u` )
Now I'm trying to implement this idea to improve sw tessellation.
I believe this idea can solve DX9's registers issues with no performance drop.

**Comment by unknownbrackets** [comment]
Ah, that's a good idea.

-[Unknown]

**Comment by xebra** [comment]
Just rebased.

**Comment by xebra** [comment]
Use PrepareBuffers() like Vulkan.

**Comment by hrydgard** [comment]
Ready for merge?

**Comment by xebra** [comment]
Please just wait, it needs compatibility check for VTF. I'll implement soon.

**Comment by xebra** [comment]
I'm not sure. Still has registers issues but it's avoiding to calculate normals for now.
So probably it is not so critical issues.
However we might need some more test before major release.
Anyway you can decide.
Of course, I will continue fix it if after merged.

**Comment by hrydgard** [comment]
OK. Yeah I'll merge this after 1.4.

File: UI/GameSettingsScreen.cpp
Comment (hrydgard) [inline UI/GameSettingsScreen.cpp]: Including this here breaks the build on non-windows platforms :) Need some #ifdef .

File: UI/GameSettingsScreen.cpp
Comment (hrydgard) [inline UI/GameSettingsScreen.cpp]: We should find some more elegant way to check this (caps on the draw context perhaps) but I'm fine with an ifdef here too for now.

**Comment by hrydgard** [comment]
@xebra if you fix this up enough to make Travis happy, I'm happy to merge.

**Comment by hrydgard** [comment]
@xebra

Will there be any further work on this soon, or should I push to 1.6?

**Comment by xebra** [comment]
@hrydgard Sorry, I was away from this project for a while, but now I'm rewriting spline/bezier logic, and it's almost done. So I can commit this and fix DX9 in a little while.

**Comment by hrydgard** [comment]
@xebra by the way, wouldn't it be possible to put the instance data in a vertex stream instead of in textures? Like so:
https://msdn.microsoft.com/en-us/library/windows/desktop/bb173349(v=vs.85).aspx

In the Vulkan backend, I've now switched from textures to shader storage buffers because it's probably even more appropriate if those are available, which they are in Vulkan but definitely not in D3D9 :)

**Comment by hrydgard** [comment]
Happening or not @xebra ? :)  Feel free to close if not, I mean, software tess works okay and not too many users are still on DX9...

**Comment by zminhquanz** [comment]
@hrydgard I think , you can remove experimental because HW Tessellation is working correctly on almost API ( DX 11 , OpenGL and Vulkan ) except DX 9

**Comment by hrydgard** [comment]
@zminhquanz , good point, I'll do that.

**Comment by xebra** [comment]
@hrydgard Sorry for my late reply, and thank you so much for your advice and caring about me.
I succeeded to implement HWtess on dx9 without errors and VTF, but I need fix some more.
Anyway, I will commit it in another push request if completed, so once closing this.

@zminhquanz My concern about "experimental" was just for old mobile devices.

**Pull request closed without merging**
